VMEncryption/volume-notif-svc/UdevVolNotif.c

```diff
@@ -0,0 +1,222 @@
+    }
+
+    struct udev *udev;
+    struct udev_monitor *mon;
+    int fd;
+    int ret;
+
+    udev = udev_new();
+    if (!udev) {
+        custom_log("ERROR: Can't create udev (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    // Set custom log function
+
+    mon = udev_monitor_new_from_netlink(udev, "udev");
+    if (!mon) {
+        custom_log("ERROR: Can't create udev monitor (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    ret = udev_monitor_filter_add_match_subsystem_devtype(mon, "block", "partition" /*"disk"*/);
+    if (ret < 0) {
+        custom_log("ERROR: Can't add udev monitor filter (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    ret = udev_monitor_enable_receiving(mon);
+    if (ret < 0) {
+        custom_log("ERROR: Can't enable udev monitor receiving (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    fd = udev_monitor_get_fd(mon);
+    if (fd < 0) {
+        custom_log("ERROR: Can't get udev monitor fd (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    custom_log("Entering main event loop");
+    while (1) {
+        fd_set fds;
+        struct timeval tv;
+        int ret;
+
+        FD_ZERO(&fds);
+        FD_SET(fd, &fds);
+        tv.tv_sec = 0;
+        tv.tv_usec = 0;
+
+        ret = select(fd+1, &fds, NULL, NULL, &tv);
+
+        if (ret > 0 && FD_ISSET(fd, &fds)) {
+            struct udev_device *dev;
+
+            dev = udev_monitor_receive_device(mon);
+            custom_log("A new udev monitoring event is received!");
+
+            const char *action = udev_device_get_action(dev);
+            custom_log("Action: %s", action);
+            const char *devtype = udev_device_get_devtype(dev);
+            custom_log("Device type: %s", devtype);
+            const char *subsystem = udev_device_get_subsystem(dev);
+            custom_log("Subsystem: %s", subsystem);
+            const char *devpath = udev_device_get_devpath(dev);
+            custom_log("Device path: %s", devpath);
+            const char *devnode = udev_device_get_devnode(dev);
+            custom_log("Device node: %s", devnode);
+            int is_initialized = udev_device_get_is_initialized(dev);
+            custom_log("is_initialized: %d", is_initialized);
+            const char *fsType = udev_device_get_property_value(dev, "ID_FS_TYPE");
+            custom_log("Filesystem type: %s", fsType);
+
+
+            if (action != NULL && strcmp(action, "change") == 0) {
+                // Handle disk volume change event
+                const char *devnode = udev_device_get_devnode(dev);
+                custom_log("Device node: %s has changed. Checking if it is encrypted!", devnode);
+
+                // TODO: Call ADE's handle sh with enable action to scan the volumes and encrypt if necessary.
+                // Detect ADE path.
+                // system("");
+            }
+            custom_log("Processing udev monitoring event is done!\n");
+            udev_device_unref(dev);
```

Comment:
multiple unref s making difficult to prove correctness better to do:
takeref and allocate resources
callfoo() new funtion
deref and deallocate resources